tasks/publish_statistics.rb

```diff
@@ -50,15 +50,18 @@
 
     logger.info("Creating #{adverbial} metrics for S3 with #{args[:date]}")
 
-    metrics = Metrics::ActiveUsers.new(period: period, date: args[:date])
+    active_users_metrics = Metrics::ActiveUsers.new(period: period, date: args[:date])
+    roaming_users_metrics = Metrics::RoadmingUsers.new(period: period, date: args[:date])
```

Comment:
If we're going to replicate the active users metrics interface, can we use inheritance instead? I feel like with the growing number of metrics, and the potential new ones we can extract from the Logging API it might be easier to go with a unified interface? Something simpler than the previous PP-code (with its substantial number of indirections) but allows new metrics to be pushed to S3 easily:

```ruby
class Metric
  VALID_PERIODS = %w[week day month].freeze

    def initialize(attrs)
      raise ArgumentError unless VALID_PERIODS.include? attrs[:period]

      @period = attrs[:period]
      @date = attrs[:date]
      @s3 = Aws::S3::Client.new(region: "eu-west-2")
    end

    def generate!
      raise "subclass must implement generate!"
    end

    def key
      raise "subclass must implement key"
    end

    def publish!
      if @report.nil?
        raise "Nothing to publish; call generate! before publishing."
      end

      bucket = ENV.fetch("S3_METRICS_BUCKET")

      @s3.put_object(
        bucket: bucket,
        key: key,
        body: report.to_json.to_s,
      )
    end    
  end
end

class RoamingUsersMetrics < Metric
  def generate!
    @report = "data"
  end

  def key
    "roaming-users/roaming-users-#{period}-#{date}"
  end
end
```

It might also make testing easier as we can focus the subclass tests on the data generation and nothing else.

Reply:
I made some changes, extracted the S3 publishing stuff to a separate class.